### DISCUSSION TITLE
FR: Wire autorouting

### DISCUSSION THREAD

**Discussion by FalcoGer** [FakeFishGames/Barotrauma#9931] [Ideas and suggestions]
I'd like a feature that automatically routes wires in the submarine editor. It could work something like this:

There is a new tool to mark areas for wiring routing similar to the hulls tool. Each area has properties that can be edited: A wireing penalty and two snap to grid checkboxes, one horizontal, one vertical. Those will snap the nodes to the grid. Any area that is not marked will have a default penalty value. Care must be taken to disallow negative valus, otherwise endless loops may result. When areas overlap, the minimum penalty value applies. Each snapping option in that case is the result of OR-ing that same snapping option from all overlapping areas. The snapping only applies to new nodes. If a machine's attachment point is not aligned, then it will stay where it is, only new nodes that are generated for the pathing will be constrained.

Wires should also have a new property to exclude them from autorouting. Autorouting applies to all wires or any one selected wire by clicking a button. A path finding algorithm, such as A*, may then be used to generate the path from one end of the wire to the other with the least amount of penalty and while being in constraints of the node snapping and the wire is laid out automatically to fit that path.

When placing cable holders, an automatic preferred area may be created with the respective snapping to grid option active.

In this example the green area forces wire nodes to be snapped horizontally, the blue area vertically. Where they overlap, the enforce both horizontal and vertical snapping to the grid. Both green and blue are a low pathing cost. Black is unassigned and has a normal pathing cost with no snapping behavior while red has a high pathing cost, forcing all entries into it to be at right angles if possible. Snapping behavior may prohibit this. If red were right up to the low cost paths, the angles would be 90°, but with the black in between, going further along a low cost path will get you less shorter travel across black for more overall path length, so the angles may vary, depending on the actual cost values.

![image](https://user-images.githubusercontent.com/16250196/191999685-9a8425f0-137e-44f1-ac4d-e83f888f2628.png)

**Comment by Saevon** [comment]
See #4713 

I'd have this in more then the sub editor btw, make wiring clean and easy in-game too
